skfda/datasets/_samples_generators.py

```diff
@@ -135,104 +255,296 @@ def euler_maruyama(  # noqa: WPS210
     """
     random_state = validate_random_state(random_state)
 
-    if n_samples is None:
-        if callable(initial_condition):
-            raise ValueError(
-                "Invalid initial conditions. If a function is given, the "
-                "n_samples argument must be included.",
-            )
+    (
+        initial_values,
+        n_samples,
+        dim_codomain,
+    ) = _sde_initial_condition_preprocessing(
+        initial_condition,
+        n_samples,
+        random_state,
+    )
 
-        initial_values = np.atleast_1d(initial_condition)
-        n_samples = len(initial_values)
-    else:
-        if callable(initial_condition):
-            initial_values = initial_condition(
-                size=n_samples,
-                random_state=random_state,
-            )
-        else:
-            initial_condition = np.atleast_1d(initial_condition)
-            dim_codomain = len(initial_condition)
-            initial_values = (
-                initial_condition
-                * np.ones((n_samples, dim_codomain))
-            )
+    if dim_codomain == 1:
+        diffusion_matricial_term = False
 
-    if initial_values.ndim == 1:
-        initial_values = initial_values[:, np.newaxis]
-    elif initial_values.ndim > 2:
-        raise ValueError(
-            "Invalid initial conditions. Each of the starting points "
-            "must be a flat array.",
+    (
+        dim_noise,
+        drift_function,
+        _,
+        diffusion_times_noise,
+    ) = _sde_drift_diffusion_preprocessing(
+        drift,
+        diffusion,
+        dim_codomain,
+        initial_values,
+        start,
+        diffusion_matricial_term,
+    )
+
+    data_matrix = np.zeros((n_samples, n_grid_points, dim_codomain))
+    times = np.linspace(start, stop, n_grid_points)
+    delta_t = times[1:] - times[:-1]
+    noise = random_state.standard_normal(
+        size=(n_samples, n_grid_points - 1, dim_noise),
+    )
+    data_matrix[:, 0] = initial_values
+
+    for n in range(n_grid_points - 1):
+        t_n = times[n]
+        x_n = data_matrix[:, n]
+
+        data_matrix[:, n + 1] = (
+            x_n
+            + delta_t[n] * drift_function(t_n, x_n)
+            + diffusion_times_noise(t_n, x_n, noise[:, n])
+            * np.sqrt(delta_t[n])
         )
-    (n_samples, dim_codomain) = initial_values.shape
 
-    if dim_codomain == 1:
-        diffusion_matricial_term = False
+    return FDataGrid(
+        grid_points=times,
+        data_matrix=data_matrix,
+    )
 
-    if drift is None:
-        drift = 0.0  # noqa: WPS358 -- Distinguish float from integer
 
-    if callable(drift):
-        drift_function = drift
-    else:
-        def constant_drift(  # noqa: WPS430 -- We need internal functions
-            t: float,
-            x: NDArrayFloat,
-        ) -> NDArrayFloat:
-            return np.atleast_1d(drift)
+def milstein(  # noqa: WPS211, WPS210
+    initial_condition: ArrayLike | InitialValueGenerator,
```

Comment:
I forgot to tell you in the previous PR, but I think that for these methods is better to require to always use keyword parameters, and no positional ones:
```suggestion
def milstein(  # noqa: WPS211, WPS210
    *,
    initial_condition: ArrayLike | InitialValueGenerator,
```